db_implementation/Offsets.ipynb

```diff
@@ -0,0 +1,369 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "1871334f-3cc0-4a64-a433-2da595a8188c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "45e84189-e7db-4699-9da7-e36425c8d899",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "f = open('mpqa_csds.json', 'r', encoding='utf-8')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "dfefc278-dcf2-4466-a981-efb40d461a80",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data = json.load(f)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 31,
+   "id": "c3fda458-b53d-490d-a333-5db5f165064b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import re\n",
+    "\n",
+    "def replace_punc(clean_text, offset, offset_list):\n",
+    "    poss_dif = 0\n",
+    "    s = clean_text\n",
+    "    s, off1 = re.subn(' \\.', '.', s)\n",
+    "    s, off2 = re.subn(' ,', ',', s)\n",
+    "    s, off3 = re.subn(' ;', ';', s)\n",
+    "    s, off4 = re.subn(' :', ':', s)\n",
+    "    s, off5 = re.subn(' POSS', '\\'s', s)\n",
+    "    if off5 > 0:\n",
+    "        poss_dif = 3\n",
+    "    adjust_val = off1+off2+off3+off4+poss_dif\n",
+    "    offset -= adjust_val\n",
+    "    if (off1+off2+off3+off4+poss_dif > 0):\n",
+    "        offset_list[-1] -= adjust_val\n",
+    "    return s, offset, offset_list"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "a3a3817f-d56e-455c-aba0-77143489ba4a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def assemble_tokens(w_text):\n",
+    "    clean_text = ''\n",
+    "    offset = 0\n",
+    "    offset_list = [0]\n",
+    "    index = 0\n",
+    "    in_quote = False\n",
```

Comment:
This logic does not seem to me to account for nested quotations.  In American usage, you are likely to encounter an open single quotation mark only after an open double quotation mark — in which case, with respect to the outer quotation, you are still "in quote."  In order to fix this problem in all its generality, you actually need a somewhat more complex stack logic.  Are you familiar with stacks?  For now, just to get through this set of changes, perhaps you can assume that there will only be at most one nesting level in quotations, and you can get by with a single in_double_quote and another in_single_quote.  Then move the handler logic to a function to avoid code duplication.